sympy/solvers/solveset.py

```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
 
 def _solve_as_rational(f, symbol, domain):
     """ solve rational functions"""
+    from sympy.core import expand_log
     f = together(f, deep=True)
     g, h = fraction(f)
+    g, h = g.expand(), h.expand()
     if not h.has(symbol):
         return _solve_as_poly(g, symbol, domain)
     else:
-        valid_solns = _solveset(g, symbol, domain)
-        invalid_solns = _solveset(h, symbol, domain)
-        return valid_solns - invalid_solns
+        # search for exp(I*x) terms
+        flag = False
+        atoms = list(g.atoms(exp))
+        atoms += list(h.atoms(exp))
+        for atom in atoms:
+            for img in expand_log(atom, force=True).atoms(I):
+                if img:
+                    flag = True
+        if flag:
+            # to get soln in desired form  for
+            # solveset((tan(x)-1).rewrite(exp), x, S.Reals)
+            # and these type of expressions, we need to
+            # make polynomial equation by replace I*symbol with dummy.
+            y = Dummy('y')
+            # make it independent from exp(I*symbol)
+            g, h = g.subs(exp(I*symbol), y), h.subs(exp(I*symbol), y)
+
+            solns = _solveset(g, y, domain) - _solveset(h, y, domain)
+            if g.has(symbol) or h.has(symbol):
+                return ConditionSet(symbol, Eq(f, 0), S.Reals)
+            if isinstance(solns, FiniteSet):
+                result = Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
+                       for s in solns])
+                return result.intersection(domain) \
+                if domain.is_subset(S.Reals) else result
+            elif solns is S.EmptySet:
+                return S.EmptySet
+            else:
+                return ConditionSet(symbol, Eq(f_original, 0), S.Reals)
+        else:
+            valid_solns = _solveset(g, symbol, domain)
+            invalid_solns = _solveset(h, symbol, domain)
+            return valid_solns - invalid_solns
 
 
 def _solve_trig(f, symbol, domain):
     """ Helper to solve trigonometric equations """
-    f = trigsimp(f)
-    f_original = f
-    f = f.rewrite(exp)
-    f = together(f)
-    g, h = fraction(f)
-    y = Dummy('y')
-    g, h = g.expand(), h.expand()
-    g, h = g.subs(exp(I*symbol), y), h.subs(exp(I*symbol), y)
-    if g.has(symbol) or h.has(symbol):
-        return ConditionSet(symbol, Eq(f, 0), S.Reals)
+    from sympy import factor, factor_list
+    from sympy.simplify.fu import hyper_as_trig, fu
 
-    solns = solveset_complex(g, y) - solveset_complex(h, y)
+    if _is_function_class_equation(TrigonometricFunction, f, symbol):
+        f = trigsimp(f)
+    # trigsimp is not defined for hyperbolic
+    elif _is_function_class_equation(HyperbolicFunction, f, symbol):
+        try:
```

Comment:
```
In [1]: f = -sinh(x)**2 + cosh(x)**2

In [2]: from sympy.simplify.fu import hyper_as_trig, fu

In [3]: t, f = hyper_as_trig(f)

In [4]: t
Out[4]: 
   2                 2           
sin (Dummy₁₈⋅x) + cos (Dummy₁₈⋅x)

In [5]: f
Out[5]: <function sympy.simplify.fu.<lambda>>

In [7]: fu(t)
Out[7]: 1  # so f in simplified to 1 
```

Mor detail is [here](https://github.com/sympy/sympy/blob/master/sympy/simplify/fu.py#L2112)